### PR TITLE
Add fallback hydration for properties

### LINKED ISSUE #156: Make changes from public/protected/private to public/protected/private safe to do so?
Emulate what php-src/ext/standard/var_unserializer.re does

This would go in 2.1.0

Related to https://github.com/igbinary/igbinary/issues/137

### LINKED ISSUE #387: igbinary_unserialize dynamically creates conflicting properties when property access has changed
This may be related to #156, I'm not sure. But that's a very old ticket so I'm wondering if I can expect a change to this behaviour, or at least a good workaround.

When a class definition has changed since serialization, it's possible for `igbinary_unserialize` to result in **multiple properties with the same name**, but different public/private access. I only noticed this behaviour when upgrading to PHP 8.2 where the creation of dynamic properties is deprecated, but the issue is reproducible in earlier PHP versions too.

The behaviour is best demonstrated with a working example:

Firstly, serialize some legacy code (a simple object with a public property) and store the binary data somewhere:

     class Foo {
        public $bar;
     }
     $legacy = new Foo;
     $legacy->bar = 'baz';
     echo igbinary_serialize($legacy); // store the output

Then update your class definition to use a private variable, and a default value to highlight the problem:

     class Foo {
        
        private string $bar = 'default';
        
        public function getBar():string {
            return $this->bar;
        }
    }

Then unserialize your saved blob. I'll do this with the standard PHP unserialize function for comparison:

    $data = 'O:3:"Foo":1:{s:3:"bar";s:3:"baz";}';
    echo "php unserialize => ",unserialize($data)->getBar(),"\n";

    $data = "\0\0\0\2\x17\3Foo\x14\1\x11\3bar\x11\3baz";
    echo "igbinary_unserialize => ",igbinary_unserialize($data)->getBar(),"\n";

The above prints as follows, showing the PHP unserialize function is unaffected, but igbinary doesn't set the private property:

    php unserialize => baz
    igbinary_unserialize => default

The latter does in fact set a property, but it dynamically sets a public property with the same name. Under PHP >= 8.2 you'll get a deprecation warning unless you annotate the class with `#[AllowDynamicProperties]`, but the behaviour exists regardless. If you dump the instance you'll see two properties named "bar" as follows:

    object(Foo)#1 (2) {
      ["bar":"Foo":private]=>
      string(7) "default"
      ["bar"]=>
      string(3) "baz"
    }

It was my own stupid fault for changing the class definition of serialized objects without a proper migration strategy, but I was blissfully unaware of this issue, and I don't imagine this is uncommon.

The only way I've found to access the originally serialized property is by casting the object to an array, like this:

    $foo = (array) igbinary_unserialize($data);
    echo "dynamic => ",$foo['bar'],"\n";
    echo "private => ",$foo["\0Foo\0bar"],"\n";

Produces:

     dynamic => baz
     private => default

 
I'm currently working around this issue using a `__wakeup` method that copies the dynamically set property from the array (as shown above) back onto the private member. However, this isn't a permanent fix as the dynamically set property still exists in duplicate and will be serialized again. I've not found any way to unset it.

It seems that in PHP 8.3 the deprecated behaviour is still not removed, but I anticipate one day it will be and all my legacy objects will not be unserializable. I know this may not be considered a "bug" as such, but are there any plans to make `igbinary_unserialize` behave as per `unserialize` in this respect?

### LINKED ISSUE #410: apcu_store()/apcu_fetch() corrupts objects that define __serialize() without __unserialize() — duplicate private+public properties
When storing an object via apcu_store() where the class defines __serialize() but not __unserialize(), the object returned by apcu_fetch() is corrupted: it contains both a private property slot (with the default value) and a dynamic public property (with the actual value) under the same name simultaneously. This state is impossible to achieve through normal PHP code.

**Environment:**
PHP version: 8.4.18
APCu version: 5.1.28
igbinary version: 3.2.16
OS: Linux Ubuntu

**APCu config:**
apc.entries_hint = 100000
apc.slam_defense = On
apc.gc_ttl = 600
apc.serializer = igbinary
apc.shm_size = 128M
apc.ttl = 3600

Code for reproducing:
```
<?php
class Foo
{
    public function __construct(private string $name = 'default') {}
    public function __serialize(): array { return ['name' => $this->name]; }
    public function getName(): string { return $this->name; }

    // __unserialize() is intentionally undefined
    // public function __unserialize(array $data): void { $this->name = $data['name']; }
}

$original = new Foo('actual');
apcu_store('test_key', $original);
$restored = apcu_fetch('test_key');

echo "Original name: " . $original->getName() . PHP_EOL;
echo "Restored name: " . $restored->getName() . PHP_EOL;

$arr = (array) $restored;
echo PHP_EOL . "Cast to array (raw keys):" . PHP_EOL;
foreach ($arr as $key => $value) {
    echo "  " . json_encode($key) . " => " . json_encode($value) . PHP_EOL;
}
```
When `apc.serializer = igbinary` enabled:
Image

After disabled: 
Image

See, two discussions about investigations this bug:
👉 https://github.com/krakjoe/apcu/issues/609
👉 https://github.com/symfony/symfony/issues/63431

### PR DESCRIPTION
**Pull request by madmajestro**
If the visibility of a property has changed between serialization and unserialization, or if a private or protected property needs to be restored, but the serialized data was generated by a __serialize method while no __unserialize method exists in the class, the property value cannot be found during hydration in the object's hash table. The reason for this is that in these cases the property name in the serialized representation does not match the prefixed/mangled property name of the current class definition. To resolve this, an attempt is made to determine the correctly prefixed property name by performing a lookup with the unprefixed property name in the properties_info hash table of the class.The implementation is borrowed from the current version of php-src/ext/standard/var_unserializer.re to align the behavior to the serializer of PHP >= 7.2. See 7cb5bdf64a95bd70623d33d6ea122c13b01113bd / https://github.com/php/php-src/pull/2494 for the initial implementation in PHP 7.2.

Closes #410
Closes #387
Closes #156

### CODE REVIEW

**Comment by stof** [comment]
I would suggest adding another test where both `__serialize` and `__unserialize` are implemented, to make sure the new logic does not break that case.

File: src/php7/igbinary.c [hunk 1]
```diff
@@ -2747,6 +2747,16 @@ inline static int igbinary_unserialize_object_properties(struct igbinary_unseria
 		return 0;
 	}
 
+#if PHP_VERSION_ID >= 70400
+	int mangle_property_names = 0;
+
+#if PHP_VERSION_ID >= 80000
+	if (ce->__serialize) mangle_property_names = 1;
```
Comment (tricky) [inline src/php7/igbinary.c]: Have not yet reviewed actual substance yet, but for style consistency; please use an explicit comparison and braces:
```
if (ce->__serialize != NULL) {
...
}
```

Also please use tabs for indentation.

**Comment by tricky** [comment]
> I would suggest adding another test where both `__serialize` and `__unserialize` are implemented, to make sure the new logic does not break that case.

Indeed this is very much required, though there are quite a few tests already. Tests can always be improved.

File: src/php7/igbinary.c (hunk 1 above)
Comment (madmajestro) [reply src/php7/igbinary.c]: I'm now using an explicit comparison and curly braces, but I've already used tabs in the C code and spaces in the PHPT tests, just like the other `__serialize*.phpt` tests do. Should I still use tabs in the tests?

**Comment by madmajestro** [comment]
I've added a test to check the behavior of `__unserialize`. Additionally, the values ​​in the `__serialize` and `__unserialize` methods are now multiplied by 10. This allows to check if the `__serialize` / `__unserialize` functions were actually used.

**Comment by madmajestro** [comment]
I have just pushed an optimized version. This version uses the mangled representation of the property name already present in zend_property_info. This saves the memory allocation and string copying done in `zend_mangle_property_name()`.

**Comment by madmajestro** [comment]
@tricky
Friendly Ping

**Comment by madmajestro** [comment]
Hm, i just saw that #387 and maybe #156 could also be resolved with this PR, if i implement the fallback logic in a different way. I will push a new version to resolve that as well...

**Comment by madmajestro** [comment]
I have now implemented the fallback hydration, as it exists in PHP's internal serializer since PHP 7.2. This should resolve several issues. In addition, the behavior of igbinary hopefully should now be identical to PHP's serializer, allowing it to be used again as a drop-in replacement.

**Comment by madmajestro** [comment]
@tricky
Friendly Ping

**Comment by madmajestro** [comment]
I've re-analyzed the behavior of the PHP serializer and Igbinary once more and noticed a few edge cases where PHP and Igbinary behaved differently. I've fixed these issues and added / revised tests so that the behavior of PHP and Igbinary is now tested and compared during the tests. I hope this simplifies the review process and helps to identify differences between PHP and Igbinary in the future.

**Comment by madmajestro** [comment]
Friendly ping

**Comment by madmajestro** [comment]
Friendly Ping

File: src/php7/igbinary.c [hunk 2]
```diff
@@ -2827,8 +2827,51 @@ inline static int igbinary_unserialize_object_properties(struct igbinary_unseria
...
+					info = zend_hash_str_find_ptr(&ce->properties_info, unmangled_prop, unmangled_prop_len);
+				}
+			}
+
+			if (info != NULL) {
+#if PHP_VERSION_ID >= 80400
+				if ((info->flags & ZEND_ACC_VIRTUAL)) {
+					php_error_docref(NULL, E_WARNING,
+						"Cannot unserialize value for virtual property %s::$%s",
+						ZSTR_VAL(info->ce->name), ZSTR_VAL(key_str));
```
Comment (andypost) [inline src/php7/igbinary.c]: When this branch fires for a private or protected virtual property, `key_str` is still the mangled `zend_string` (e.g. `\0Class\0prop` for private, `\0*\0prop` for protected). `ZSTR_VAL(key_str)` then contains embedded NULs, and `%s` in `php_error_docref` truncates at the first NUL — the user sees `Cannot unserialize value for virtual property Test::$` followed by garbage/nothing instead of the property name.

The included `__serialize_025.phpt` only exercises a public virtual property, so this case is not covered by tests. Suggest using `info->name` (the unmangled name from `properties_info`) or the local `unmangled_prop`:

```c
"Cannot unserialize value for virtual property %s::$%s",
ZSTR_VAL(info->ce->name), ZSTR_VAL(info->name));
```

**Comment by andypost** [comment]
Stylistic nit, not blocking — a few lines below the new fallback block, the existing code reassigns `info` via `zend_get_typed_property_info_for_slot(...)` (around `src/php7/igbinary.c:2881`):

```c
info = zend_get_typed_property_info_for_slot(Z_OBJ_P(z_deref), prototype_value);
```

When the fallback above populated `info` from `ce->properties_info`, that value is unconditionally overwritten here. Functionally correct (the type-source registration below needs the typed-property variant), but it is easy to miss that the fallback `info` is only used to recover the property *name* and never flows past this point. A one-line comment like `/* override fallback info; below we need the typed-property variant */` next to that assignment would save the next reader a minute of head-scratching.

File: src/php7/igbinary.c (hunk 2 above)
Comment (andypost) [reply src/php7/igbinary.c]: The test to add
```
--TEST--
Virtual-property warning is truncated when key was originally protected/private
--SKIPIF--
<?php if (PHP_VERSION_ID < 80400) { echo "skip virtual properties need PHP 8.4+"; } ?>
--FILE--
<?php
// Bytes generated once (PHP 8.5) from:
//     class Test { protected int $prop = 1; }
//     bin2hex(igbinary_serialize(new Test()));
// 17 04 "Test" | 14 01 (array,1) | 11 07 \0*\0prop | 06 01 (int 1)
$data = hex2bin('0000000217045465737414011107002a0070726f700601');

// Same class name, but `prop` is now a public *virtual* property.
class Test {
    public int $prop { 
        get => 1;
    }
}

igbinary_unserialize($data);
?>
--EXPECTF--
Warning: igbinary_unserialize(): Cannot unserialize value for virtual property Test::$prop in %s on line %d
```

Comment (madmajestro) [reply src/php7/igbinary.c]: Thanks for suggesting a fix and a test!

**Comment by madmajestro** [comment]
> Stylistic nit, not blocking — a few lines below the new fallback block, the existing code reassigns `info` via `zend_get_typed_property_info_for_slot(...)` (around `src/php7/igbinary.c:2881`):

I am now using two different variables. That makes it clearer.

**Comment by dshafik** [comment]
This issue is blocking our upgrade to PHP 8.5, is there anything blocking this being merged and a final 3.2.17 release being created?
